pandas/core/strings.py

```diff
@@ -1742,12 +1745,80 @@ def str_encode(arr, encoding, errors="strict"):
+    However, different methods allow different input types, and so this can not
+    be checked during :meth:`StringMethods.__init__`, but must be done on a
+    per-method basis. This decorator exists to facilitate this process, and
+    make it explicit which (inferred) types are disallowed by the method.
+
+    :meth:`StringMethods.__init__` allows the *union* of types its different
+    methods allow (after skipping NaNs; see :meth:`StringMethods._validate`):
+    ['string', 'unicode', 'empty', 'bytes', 'mixed', 'mixed-integer'].
+
+    The default string types ['string', 'unicode', 'empty'] are allowed for all
+    methods. For the additional types ['bytes', 'mixed', 'mixed-integer'], each
+    method then needs to forbid the types it is not intended for.
+
+    Parameters
+    ----------
+    forbidden : list or None
+        List of forbidden non-string types, may be one or more of
+        `['bytes', 'mixed', 'mixed-integer']`.
+    name : str, default None
+        Name of the method to use in the error message. By default, this is
+        None, in which case the name from the method being wrapped will be
+        copied. However, for working with further wrappers (like _pat_wrapper
+        and _noarg_wrapper), it is necessary to specify the name.
+
+    Returns
+    -------
+    func : wrapper
+        The method to which the decorator is applied, with an added check that
+        enforces the inferred type to not be in the list of forbidden types.
+
+    Raises
+    ------
+    AttributeError
+        If the inferred type of the underlying data is in `forbidden`.
+    """
+
+    # deal with None
+    forbidden = [] if forbidden is None else forbidden
+    # deal with single string instead of list
+    forbidden = [forbidden] if isinstance(forbidden, str) else forbidden
+
+    allowed_types = {'string', 'unicode', 'empty',
+                     'bytes', 'mixed', 'mixed-integer'} - set(forbidden)
+
+    def _forbid_nonstring_types(func):
+        func_name = func.__name__ if name is None else name
+
+        @wraps(func)
+        def wrapper(self, *args, **kwargs):
+            if self._inferred_type not in allowed_types:
+                msg = ("Cannot use .str.{name} with values of inferred type "
+                       "{inf_type!r}.".format(name=func_name,
+                                              inf_type=self._inferred_type))
+                raise AttributeError(msg)
+            return func(self, *args, **kwargs)
+        wrapper.__name__ = func_name
+        return wrapper
+    return _forbid_nonstring_types
+
+
+def _noarg_wrapper(f, name=None, docstring=None, forbidden_types=['bytes'],
+                   **kargs):
+    @forbid_nonstring_types(forbidden_types, name=name)
     def wrapper(self):
         result = _na_map(f, self._parent, **kargs)
         return self._wrap_result(result)
 
-    wrapper.__name__ = f.__name__
+    wrapper.__name__ = f.__name__ if name is None else name
```

Comment:
These names were wrong (just showing as `<lambda>`, and were causing test fails, because the error message of `forbid_nonstring_types` is method-specific.

Reply:
Now added an issue reference: #23551. I thought I could split this off relatively easily, but then I'd have to rebuild the whole parametrized method-fixture just for checking the names.